cl/users/email_handlers.py

```diff
@@ -481,3 +357,153 @@ def add_bcc_random(
+    # retry messages as soon as possible.
+    return now()
+
+
+def is_message_stored(message_id: str) -> tuple[bool, int | None]:
+    """Returns True if the message is stored in database.
+
+    :param message_id: The message unique identifier.
+    :return bool: True if the message is available in database, otherwise False
+    """
+    if message_id:
+        stored_email = EmailSent.objects.filter(message_id=message_id)
+        if stored_email.exists():
+            return True, stored_email[0].pk
+    return False, None
+
+
+from cl.users.tasks import send_failed_email
+
+
+def enqueue_email(recipients: list[str], message_id: str) -> None:
+    """Enqueue a message for a list of recipients, due to a soft bounce or if
+    the recipient is under a backoff event waiting period.
+
+    :param recipients: The list of recipients to enqueue the message.
+    :param message_id: The message unique identifier.
+    :return None:
+    """
+
+    stored, stored_email_pk = is_message_stored(message_id)
+    if stored:
+        for recipient in recipients:
+            # Before clean possible failed task objects.
+            delete_failed_tasks_objects(recipient)
+
+            # Get the next retry datetime to schedule the message, based on the
+            # active backoff event.
+            scheduled_datetime = get_next_retry_date(recipient)
+
+            # If not exists here we create the FailedEmail object that will
+            # help us try to unblock the user's email address once the backoff
+            # event waiting period expires, status: ENQUEUED
+            failed, created = FailedEmail.objects.get_or_create(
+                recipient=recipient,
+                status=STATUS_TYPES.ENQUEUED,
+                defaults={
+                    "stored_email_id": stored_email_pk,
+                    "next_retry_date": scheduled_datetime,
+                },
+            )
+            if created:
+                # If the ENQUEUED FailedEmail object was created, schedule the
+                # celery task, in this way we ensure we have a message to send
+                # to the user once the backoff event waiting period expires
+                send_failed_email(failed.pk, scheduled_datetime)
+            else:
+                # If the recipient has already one ENQUEUED FailedEmail the
+                # following messages will be enqueued with WAITING status.
+                # These  objects are going to be scheduled when receiving a
+                # delivery notification.
+                FailedEmail.objects.create(
+                    recipient=recipient,
+                    status=STATUS_TYPES.WAITING,
+                    stored_email_id=stored_email_pk,
+                    next_retry_date=scheduled_datetime,
+                )
+    else:
+        logging.warning(
+            f"The message: {message_id} can't be enqueued because it "
+            "doesn't exist anymore."
+        )
```

Comment:
I think I'd put this in an if at the top, and then you wouldn't need an else, and everything could be outdented.

Reply:
Changed!